exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/auth/DrillSpnegoAuthenticator.java

```diff
@@ -20,165 +20,142 @@
 import org.eclipse.jetty.server.Request;
-import org.eclipse.jetty.server.UserIdentity;
+import org.eclipse.jetty.server.Response;
+import org.eclipse.jetty.util.Callback;
 
-import javax.servlet.ServletRequest;
-import javax.servlet.ServletResponse;
-import javax.servlet.http.HttpServletRequest;
-import javax.servlet.http.HttpServletResponse;
-import javax.servlet.http.HttpSession;
-import java.io.IOException;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpSession;
 
 /**
- * Custom SpnegoAuthenticator for Drill
+ * Custom SpnegoAuthenticator for Drill - Jetty 12 version
+ *
+ * This class extends LoginAuthenticator and provides SPNEGO authentication support.
  */
-public class DrillSpnegoAuthenticator extends SpnegoAuthenticator {
+public class DrillSpnegoAuthenticator extends LoginAuthenticator {
 
   private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(DrillSpnegoAuthenticator.class);
+  private static final String AUTH_METHOD = "SPNEGO";
 
-  public DrillSpnegoAuthenticator(String authMethod) {
-    super(authMethod);
+  public DrillSpnegoAuthenticator() {
+    super();
   }
 
+
   /**
-   * Updated logic as compared to default implementation in
-   * {@link SpnegoAuthenticator#validateRequest(ServletRequest, ServletResponse, boolean)} to handle below cases:
-   * 1) Perform SPNEGO authentication only when spnegoLogin resource is requested. This helps to avoid authentication
-   *    for each and every resource which the JETTY provided authenticator does.
-   * 2) Helps to redirect to the target URL after authentication is done successfully.
-   * 3) Clear-Up in memory session information once LogOut is triggered such that any future request also triggers SPNEGO
-   *    authentication.
-   * @param request
-   * @param response
-   * @param mandatoryAuth
-   * @return
-   * @throws ServerAuthException
+   * Jetty 12 validateRequest implementation using core Request/Response/Callback API.
+   * Handles:
+   * 1) Perform SPNEGO authentication only when spnegoLogin resource is requested
+   * 2) Redirect to target URL after authentication
+   * 3) Clear session information on logout
    */
   @Override
-  public Authentication validateRequest(ServletRequest request, ServletResponse response, boolean mandatoryAuth)
+  public AuthenticationState validateRequest(Request request, Response response, Callback callback)
       throws ServerAuthException {
 
-    final HttpServletRequest req = (HttpServletRequest) request;
-    final HttpSession session = req.getSession(true);
-    final Authentication authentication = (Authentication) session.getAttribute(SessionAuthentication.__J_AUTHENTICATED);
-    final String uri = req.getRequestURI();
+    // Get the servlet request from the core request
+    ServletContextRequest servletContextRequest = Request.as(request, ServletContextRequest.class);
+    if (servletContextRequest == null) {
+      return AuthenticationState.CHALLENGE;
```

Comment:
Returning CHALLENGE requires you to trigger the challenge.
This also means you have to handle the `callback` properly.

See the implementation at `org.eclipse.jetty.security.authentication.SPNEGOAuthenticator` and it's `sendChallenge` method.